### DISCUSSION TITLE
GLSPActionDispatcher inject error after update to 2.0

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#1160] [Q&A]
hi,
   I have a custom select service in my custom tasklist-theia module. 
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/39e2d268-317e-494d-a783-48375a98f566)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/842b236e-20e2-4a3f-95db-4a2474a33e6f)
When I try to open my custom file I get the following errors
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/5d73d6d1-45da-4f14-bb61-14f06eead28e)
If I remove the @inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher in my custom select service, it is ok,
 could I not inject GLSPActionDispatcher in version 2.0?

**Comment by tortmayr** [comment]
That's interesting, looks like somehow we have introduced a circular dependency to the `ActionDispatcher` for `GLSPSelectionDataService` with  2.0. This was not intentional and has to be investigated. In the meantime you can work around this by injecting the async `ActionDispatcherProvider` instead:

```ts
@injectable()
export class TestSelectionDataService implements GlspSelectionDataService {
    protected actionDispatcher: GLSPActionDispatcher;

    constructor(@inject(TYPES.IActionDispatcherProvider) actionDispatcherProvider: () => Promise<GLSPActionDispatcher>) {
        actionDispatcherProvider().then(actionDispatcher => (this.actionDispatcher = actionDispatcher));
    }

    async getSelectionData(selectedElementIds: string[]): Promise<GlspSelectionData> {
        // do something
    }
}
```

**Comment by wss29** [comment]
hi,
    although it can get  actionDispatcher  in constructor , still has another error 'Maximum call stack size exceeded'
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/525206b3-9592-4ebb-981d-6c71be6a1b50)

**Reply by tortmayr** [reply]
@wss29 You are right. I get the same "Maximum call stack error".
  A [PR](https://github.com/eclipse-glsp/glsp-client/pull/305) with a bug fix has been opened. This problem should be resolved with the next minor Release (2.1.0) of GLSP.
 
 In the meantime, you can use the following workaround:
 Create a factory binding for the action id dispatcher. This allows you to create an instance of the dispatcher at runtime,
 which decouples it from the injection cycle and should avoid the circular dependency issue.
 
 ```ts
    export const ActionDispatcherFactory= Symbol("ActionDispatcherFactory");
    // in your custom tasklist-theia module
    bind(ActionDispatcherFactory).toFactory<GLSPActionDispatcher>(
        ctx => () => ctx.container.get<GLSPActionDispatcher>(TYPES.IActionDispatcher)
    );
    
    // In your SelectionDataService
    
    @injectable()
    export class TestSelectionDataService implements GlspSelectionDataService {
      @inject(ActionDispatcherFactory)
      protected actionDispatcherFactory: () => GLSPActionDispatcher;
    
      private _actionDispatcher: GLSPActionDispatcher | undefined;
      get actionDispatcher(): ActionDispatcher {
          if (!this._actionDispatcher) {
              this._actionDispatcher = this.actionDispatcherFactory();
          }
          return this._actionDispatcher;
      }
      
      async getSelectionData(selectedElementIds: string[]): Promise<GlspSelectionData> {
          // do something e.g:
          // this.actionDispatcher.dispatch(FooAction.create())`
      }
    }
    
    
 
 ```
